third_party/txt/src/skia/paragraph_builder_skia.cc

```diff
@@ -42,10 +42,66 @@ SkFontStyle MakeSkFontStyle(txt::FontWeight font_weight,
                                            : SkFontStyle::Slant::kItalic_Slant);
 }
 
-skt::ParagraphStyle TxtToSkia(const ParagraphStyle& txt) {
+}  // anonymous namespace
+
+ParagraphBuilderSkia::ParagraphBuilderSkia(
+    const ParagraphStyle& style,
+    std::shared_ptr<FontCollection> font_collection)
+    : base_style_(style.GetTextStyle()) {
+  builder_ = skt::ParagraphBuilder::make(
+      TxtToSkia(style), font_collection->CreateSktFontCollection());
+}
+
+ParagraphBuilderSkia::~ParagraphBuilderSkia() = default;
+
+void ParagraphBuilderSkia::PushStyle(const TextStyle& style) {
+  builder_->pushStyle(TxtToSkia(style));
+  txt_style_stack_.push(style);
+}
+
+void ParagraphBuilderSkia::Pop() {
+  builder_->pop();
+  txt_style_stack_.pop();
+}
+
+const TextStyle& ParagraphBuilderSkia::PeekStyle() {
+  return txt_style_stack_.empty() ? base_style_ : txt_style_stack_.top();
+}
+
+void ParagraphBuilderSkia::AddText(const std::u16string& text) {
+  builder_->addText(text);
+}
+
+void ParagraphBuilderSkia::AddPlaceholder(PlaceholderRun& span) {
+  skt::PlaceholderStyle placeholder_style;
+  placeholder_style.fHeight = span.height;
+  placeholder_style.fWidth = span.width;
+  placeholder_style.fBaseline = static_cast<skt::TextBaseline>(span.baseline);
+  placeholder_style.fBaselineOffset = span.baseline_offset;
+  placeholder_style.fAlignment =
+      static_cast<skt::PlaceholderAlignment>(span.alignment);
+
+  builder_->addPlaceholder(placeholder_style);
+}
+
+std::unique_ptr<Paragraph> ParagraphBuilderSkia::Build() {
+  return std::make_unique<ParagraphSkia>(builder_->Build(),
+                                         std::move(dl_paints_));
+}
+
+skt::ParagraphPainter::PaintID ParagraphBuilderSkia::CreatePaintID(
+    const flutter::DlPaint& dl_paint) {
+  dl_paints_.push_back(dl_paint);
+  return dl_paints_.size() - 1;
+}
+
+skt::ParagraphStyle ParagraphBuilderSkia::TxtToSkia(const ParagraphStyle& txt) {
   skt::ParagraphStyle skia;
   skt::TextStyle text_style;
 
+  flutter::DlPaint dl_paint;
+  dl_paint.setColor(text_style.getColor());
```

Comment:
This is odd, the color is coming from the uninitialized text_style rather than from any argument to the method. Is that right?

Reply:
Yes - all painting colors now need to be represented as `DlPaint` objects.  So this needs to take the SkParagraph `TextStyle` default color and wrap it in a `DlPaint`.

Added a comment.